(SUP-2906) Determines if Thundering Herd is occurring

asselvakumar opened the pull request:
## Please check off the steps below as you complete each step
- [ ] Put the Jira ticket number in parentheses in the **Title** e.g. `(SUP-XXXX) Add Super Duper State Check`
- [ ] Update the Jira ticket status to `Ready for Review`
- [ ] Update the Scoring Spreadsheet
- [ ] Get a **review** from someone on the support team
- [ ] Get a **review** from @misseuropa, @davidbastedo, or @deleon365

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/puppetlabs/puppetlabs-self_service?pullRequest=54) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/puppetlabs/puppetlabs-self_service?pullRequest=54) before we can accept your contribution.<br/><hr/>**Selvakumar.subburaj** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/puppetlabs/puppetlabs-self_service?pullRequest=54) it.</sub>

asselvakumar commented:
Discussed with Marty about the spec test, which am currently unable to do but I have done the functional test in my local environment by creating custom facts and the results are shown in the below screenshot.

![image](https://user-images.githubusercontent.com/37397672/151533536-7a3a6593-8891-4045-974b-2560010034e9.png)

MartyEwings commented on lib/facter/self_service.rb:
```diff
@@ -188,4 +188,19 @@
+    if differences >= 75
+      thundering_occured = true
+    end
+    { S0027: thundering_occured }
```
you can remove the if and the thundering_occured variable and change the line to 

```suggestion
    { S0027: differences >= 75 }
```
and it would have the same result with less code

You also need to flip the logic of the output here, you need this to return True when there is no herd, and false when there is a herd (true indicating  everything is OK)

There are a few ways to do this, like changing it to <= or putting a ! (NOT) in front of the expression

MartyEwings commented on lib/facter/self_service.rb:
```diff
@@ -188,4 +188,19 @@
+    thundering_occured = false
+    minvalue = thundering_query_output.min.to_i
+    maxvalue = thundering_query_output.max.to_i
+    differences = ((maxvalue - minvalue) * 100) / minvalue
```
when minvalue = 0 this crashes:

` Error while resolving custom fact fact='self_service', resolution='<anonymous>': divided by 0`

You need to account for systems with low activity

asselvakumar replied:
I did handle this minvalue=0 case, The statement below `thundering_query_output.delete(0)` helps to remove the 0 from the output data.

```
    thundering_query_output.delete(0)
    thundering_occured = false
    minvalue = thundering_query_output.min.to_i
```

asselvakumar replied:
does this confine statement correct? `next unless PuppetSelfService.primary?` i see from one of my agent nodes that the facter is executed and getting `Error: Facter: error while resolving custom facts in /opt/puppetlabs/puppet/cache/lib/facter/self_service.rb: divided by 0`


```
chunk(:S0027) do
    next unless PuppetSelfService.primary?
    # Check if thundering herd is occuring.
    thundering_query_output = PuppetSelfService.psql_thundering_herd.split(' ').drop(2).map(&:to_i)
    thundering_query_output.delete(0)
    thundering_occured = false
    minvalue = thundering_query_output.min.to_i
    maxvalue = thundering_query_output.max.to_i
    differences = ((maxvalue - minvalue) * 100) / minvalue
    if differences >= 75
      thundering_occured = true
    end
    { S0027: thundering_occured }
  end
```

MartyEwings replied:
@asselvakumar this strange, as the whole fact is confined to just infrastructure nodes. 

   next unless PuppetSelfService.primary? is a further refinement per chunk, but the Whole fact is confined by:

 ` confine { Facter.value(:pe_build) }` which is never true on an normal agent

MartyEwings replied:
> I did handle this minvalue=0 case, The statement below `thundering_query_output.delete(0)` helps to remove the 0 from the output data.
> 
> ```
>     thundering_query_output.delete(0)
>     thundering_occured = false
>     minvalue = thundering_query_output.min.to_i
> ```

The acceptance tests are still failing for a divide by zero error, which implies there are still 0's leaking into the function

MartyEwings reviewed (changes requested):
Few changes needed, the biggest is the divide by zero issue, you need to account for low activity systems in your calculations so we don't ever divide by 0

The next is some syntax changes, and the need to logic flip the output

jarretlavallee commented on lib/shared/puppet_self_service.rb:
```diff
@@ -121,4 +126,28 @@ def self.filesystem_free(path)
+    sql = %(
+      SELECT count(*)
+      FROM reports
+      WHERE start_time BETWEEN now() - interval '7 days' AND now()
```
7 days seems like a lot of data to be capturing for a test that runs every 30 minutes.

misseuropa reviewed (approved):
Has a readme table row for S0027 with a link, technically you can cut it down to https://support.puppet.com/hc/en-us/articles/215729277 but it works the way it is now.

MartyEwings reviewed (changes requested):
HI Kumar, this looks good, you have  a few merge conflicts preventing the tests running, have you put any thought into my suggestion of determining a herd?

Assuming we can account for the , wouldn't the following pattern trigger the alert?

1 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4

This is not true herd we need to worry about when dealing with small numbers, but the math would create an alert, we need a second layer of logic,

Find the average count per min over the last 30 mins,
Find any siginificantly above average
Compare this outlyer to total active nodes or a threshold and determine if this is a cluster worth worrying about.

asselvakumar commented:
> HI Kumar, this looks good, you have a few merge conflicts preventing the tests running, have you put any thought into my suggestion of determining a herd?
> 
> Assuming we can account for the , wouldn't the following pattern trigger the alert?
> 
> 1 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4
> 
> This is not true herd we need to worry about when dealing with small numbers, but the math would create an alert, we need a second layer of logic,
> 
> Find the average count per min over the last 30 mins, Find any siginificantly above average Compare this outlyer to total active nodes or a threshold and determine if this is a cluster worth worrying about.

Updated the code to handle divide by zero and changed the logic based on the standard deviation approach. Please review and let me know if any changes are required.

MartyEwings commented on lib/facter/self_service.rb:
```diff
@@ -188,4 +188,36 @@
+        # not meeting minimum requirement
+        { S0027: true }
+      else
+        # minvalue = thundering_query_output.min.to_i
+        # maxvalue = thundering_query_output.max.to_i
+        # differences = ((maxvalue - minvalue) * 100) / minvalue
+        # { S0027: differences <= 75 }
+
+        # Changing the approach based on standard deviation method
+        # taking only the first n number of items to check if thundering herd is occured or not.
```
We need to clean up the comments and older code iterations from the PR
Comments should be relevant to the end user or future developer

MartyEwings commented on spec/acceptance/self_service_spec.rb:
```diff
@@ -15,7 +15,7 @@
     describe 'check no self_service fact is false' do
       it 'if idempotent all facts should be true' do
-        expect(host_inventory['facter']['self_service'].size).to eq(20)
+        expect(host_inventory['facter']['self_service'].size).to eq(21)
```
this will need updated to the latest amount

MartyEwings commented on spec/acceptance/self_service_spec.rb:
```diff
@@ -235,6 +235,10 @@ class {'self_service':
         run_shell('systemctl start puppet_system_processes-metrics.timer')
         run_shell('puppet agent --enable')
       end
+      it 'if S0027 conditions for false are met' do
+        result = run_shell('facter -p self_service.S0027')
+        expect(result.stdout).to match(%r{false})
+      end
```
this acceptance test does not have failure conditions so all acceptance will fail, Jarret had a suggestion, but it would be cleaner not to test than to have a failure test that can not be triggered

MartyEwings reviewed (changes requested):
There is some cleanup and testing concerns that need completed.
And the Pr would need refactored to consider the name change since version 1

Good work

MartyEwings commented:
@asselvakumar is this something you have time to work on?
